### PR TITLE
[core] Avoid deprecated `std::iterator`

### PR DESCRIPTION
**Pull request by ikabadzhov**
Deprecated `std::iterator` since C++17 is substituted in:
* `TViewPubFunctionsIter`
* `TBtreeIter`
* `TListIter`
* `TMapIter`
* `TObjArrayIter`
* `TOrdCollectionIter`
* `TRefArrayIter`
* `RIndexIter`

### CODE REVIEW

**Comment by ikabadzhov** [comment]
I am working on it. PR coming tomorrow.

File: core/cont/inc/TBtree.h [hunk 1]
```diff
@@ -331,12 +331,15 @@ friend class  TBtInnerNode;
...
-                                        TObject*, std::ptrdiff_t,
-                                        const TObject**, const TObject*&> {
+class TBtreeIter : public TIterator {
+
+   using iterator_category = std::bidirectional_iterator_tag;
+   using value_type = TObject *;
+   using difference_type = std::ptrdiff_t;
+   using pointer = TObject **;
+   using const_pointer = const TObject **;
+   using reference = const TObject *&;
```
Comment (pcanal) [inline core/cont/inc/TBtree.h:341]:
```
public:
   using iterator_category = std::bidirectional_iterator_tag;
   using value_type = TObject *;
   using difference_type = std::ptrdiff_t;
   using pointer = TObject **;
   using const_pointer = const TObject **;
   using reference = const TObject *&;
private:
```

File: core/base/v7/inc/ROOT/RIndexIter.hxx [hunk 2]
```diff
@@ -32,13 +32,20 @@ namespace Internal {
 
 template <class REFERENCE,
           class POINTER = typename std::add_pointer<typename std::remove_reference<REFERENCE>::type>::type>
-class RIndexIter: public std::iterator<std::random_access_iterator_tag, REFERENCE, POINTER> {
+class RIndexIter {
```
Comment (ikabadzhov) [inline core/base/v7/inc/ROOT/RIndexIter.hxx:35]: @pcanal Could you please take a look. I simply "mocked" the current behaviour, but I suspect is wrong. The iterator was given as:
```cpp
template <class REFERENCE,
          class POINTER = typename std::add_pointer<typename std::remove_reference<REFERENCE>::type>::type>
class RIndexIter: public std::iterator<std::random_access_iterator_tag, REFERENCE, POINTER>
```
which implied that the `value_type` is of type `REFERENCE` and the `difference_type` is of type `POINTER`. Was that intentional? Or the difference is a `REFERENCE` as well?

File: core/cont/inc/TObjArray.h [hunk 3]
```diff
@@ -134,6 +131,13 @@ class TObjArrayIter : public TIterator,
    TObjArrayIter() : fArray(0), fCurCursor(0), fCursor(0), fDirection(kIterForward) { }
 
 public:
+   using iterator_category = std::bidirectional_iterator_tag; // TODO: ideally it should be a randomaccess_iterator_tag
```
Comment (ikabadzhov) [inline core/cont/inc/TObjArray.h:134]: @pcanal even the original code had this `TODO` comment. My argument here is: why going for `randomaccess_iterator_tag` when `bidirectional_iterator_tag` already does its job. Or should I change it?

Same remark applies for `TRefArrayIter`

File: core/base/v7/inc/ROOT/RIndexIter.hxx (hunk 2 above)
Comment (linev) [reply core/base/v7/inc/ROOT/RIndexIter.hxx:35]: In my mind, `REFERENCE` should be a reference and `POINTER` as pointer like:
```
   using iterator_category = std::random_access_iterator_tag;
   using value_type = typename std::remove_reference<REFERENCE>::type;
   using difference_type = std::ptrdiff_t;
   using pointer = POINTER;
   using reference = REFERENCE;
```
But it is good question for @Axel-Naumann.

Comment (linev) [reply core/base/v7/inc/ROOT/RIndexIter.hxx:35]: Just looking how `RIndexIter` used later in `RHistBinIter.hxx`

**Comment by linev** [comment]
Also, one can remove obsolete defines like:
```
#if (__GNUC__ >= 3) && !defined(__INTEL_COMPILER)
// Prevent -Weffc++ from complaining about the inheritance
// TListIter from std::iterator.
#pragma GCC system_header
#endif
```
while we no longer using inheritance from `std::iterator`

File: core/base/v7/inc/ROOT/RIndexIter.hxx (hunk 2 above)
Comment (Axel-Naumann) [reply core/base/v7/inc/ROOT/RIndexIter.hxx:35]: I agree it's weird. It's also likely to vanish once @hahnjo finds the time to re-implement views (and with it likely iteration). So I propose you just convert it 1:1 as you did, thanks!

File: core/cont/inc/TObjArray.h (hunk 3 above)
Comment (linev) [reply core/cont/inc/TObjArray.h:128]: I guess, for `randomaccess_iterator_tag` one should support more functionality - like arithmetic with iterators `begin()+5`.
It is not provided by base `TCollection` class, therefore one need more changes to get it working with `TObjArray` properly. I guess, this PR should focus on fixing compiler warnings

Comment (ikabadzhov) [reply core/cont/inc/TObjArray.h:128]: Then this PR should already be ready to go :)

**Comment by ikabadzhov** [comment]
@phsft-bot build just on mac1015/python3

**Review by pcanal — APPROVED** [review]
Thanks.

**Pull request merged**